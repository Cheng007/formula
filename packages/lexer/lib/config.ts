import type { IConfig } from "./types";
import { TOKEN_TYPE, State } from './types'

export const SPACE = ' '
export const NEW_LINE = ['\n', '\t', '\r']

const config: IConfig = {
  initialState: State.INITIAL,
  states: {
    [State.INITIAL]: {
      isEndable: false,
      transition: [
        {
          state: State.NUMBER,
          checker: /[0-9]/
        },
        {
          state: State.START_STRING,
          checker: "'"
        },
        {
          state: State.IDENTIFIER,
          checker: /[a-zA-Z]/
        },
        {
          state: State.LEFT_PAREN,
          checker: '('
        },
        {
          state: State.RIGHT_PAREN,
          checker: ')'
        },
        {
          state: State.LEFT_SQUARE_BRACKET,
          checker: '['
        },
        {
          state: State.RIGHT_SQUARE_BRACKET,
          checker: ']'
        },
        {
          state: State.LEFT_CURLY_BRACE,
          checker: '{'
        },
        {
          state: State.RIGHT_CURLY_BRACE,
          checker: '}'
        },
        {
          state: State.SEMI_COLON,
          checker: ';'
        },
        {
          state: State.COMMA,
          checker: ','
        },
        {
          state: State.DOT,
          checker: '.'
        },
        {
          state: State.COLON,
          checker: ':'
        },
        {
          state: State.PLUS,
          checker: '+'
        },
        {
          state: State.MINUS,
          checker: '-'
        },
        {
          state: State.MULTIPLY,
          checker: '*'
        },
        {
          state: State.DIVIDE,
          checker: '/'
        },
        {
          state: State.ASSIGN,
          checker: '='
        },
        {
          state: State.LESS_THAN,
          checker: '<'
        },
        {
          state: State.GREATER_THAN,
          checker: '>'
        },
      ]
    },
    [State.NUMBER]: {
      isEndable: true,
      tokenType: TOKEN_TYPE.NUMBER,
      transition: [
        {
          state: State.NUMBER,
          checker: /[0-9\.]/
        }
      ]
    },
    [State.START_STRING]: {
      isEndable: false,
      transition: [
        {
          state: State.START_STRING,
          checker: /[^']/
        },
        {
          state: State.STRING,
          checker: "'"
        }
      ]
    },
    [State.STRING]: {
      isEndable: true,
      tokenType: TOKEN_TYPE.STRING,
    },
    [State.START_IDENTIFIER]: {
      isEndable: true,
      transition: [
        {
          state: State.IDENTIFIER,
          checker: /[a-zA-Z_0-9]/
        }
      ]
    },
    [State.IDENTIFIER]: {
      isEndable: true,
      tokenType: TOKEN_TYPE.IDENTIFIER,
      transition: [
        {
          state: State.IDENTIFIER,
          checker: /[a-zA-Z]/
        }
      ]
    },
    [State.LEFT_PAREN]: {
      isEndable: true,
      tokenType: State.LEFT_PAREN,
    },
    [State.RIGHT_PAREN]: {
      isEndable: true,
      tokenType: State.RIGHT_PAREN,
    },
    [State.LEFT_SQUARE_BRACKET]: {
      isEndable: true,
      tokenType: State.LEFT_SQUARE_BRACKET,
    },
    [State.RIGHT_SQUARE_BRACKET]: {
      isEndable: true,
      tokenType: State.RIGHT_SQUARE_BRACKET,
    },
    [State.LEFT_CURLY_BRACE]: {
      isEndable: true,
      tokenType: State.LEFT_CURLY_BRACE,
    },
    [State.RIGHT_CURLY_BRACE]: {
      isEndable: true,
      tokenType: State.RIGHT_CURLY_BRACE,
    },
    [State.SEMI_COLON]: {
      isEndable: true,
      tokenType: State.SEMI_COLON,
    },
    [State.COMMA]: {
      isEndable: true,
      tokenType: State.COMMA,
    },
    [State.DOT]: {
      isEndable: true,
      tokenType: State.DOT,
    },
    [State.COLON]: {
      isEndable: true,
      tokenType: State.COLON,
    },

    [State.PLUS]: {
      isEndable: true,
      tokenType: State.PLUS,
      transition: [
        {
          state: State.PLUS_PLUS,
          checker: '+'
        },
        {
          state: State.PLUS_ASSIGN,
          checker: '=',
        }
      ]
    },
    [State.PLUS_PLUS]: {
      isEndable: true,
      tokenType: State.PLUS_PLUS,
    },
    [State.MINUS]: {
      isEndable: true,
      tokenType: State.MINUS,
      transition: [
        {
          state: State.MINUS_MINUS,
          checker: '-'
        },
        {
          state: State.MINUS_ASSIGN,
          checker: '=',
        }
      ]
    },
    [State.MINUS_MINUS]: {
      isEndable: true,
      tokenType: State.MINUS_MINUS,
    },
    [State.MULTIPLY]: {
      isEndable: true,
      tokenType: State.MULTIPLY,
      transition: [
        {
          state: State.MULTIPLY_ASSIGN,
          checker: '='
        }
      ]
    },
    [State.DIVIDE]: {
      isEndable: true,
      tokenType: State.DIVIDE,
      transition: [
        {
          state: State.DIVIDE_ASSIGN,
          checker: '='
        },
        {
          state: State.LINE_COMMENT,
          checker: '/'
        },
        {
          state: State.START_BLOCK_COMMENT,
          checker: '*'
        }
      ]
    },
    [State.PLUS_ASSIGN]: {
      isEndable: true,
      tokenType: State.PLUS_ASSIGN,
    },
    [State.MINUS_ASSIGN]: {
      isEndable: true,
      tokenType: State.MINUS_ASSIGN,
    },
    [State.MULTIPLY_ASSIGN]: {
      isEndable: true,
      tokenType: State.MULTIPLY_ASSIGN,
    },
    [State.DIVIDE_ASSIGN]: {
      isEndable: true,
      tokenType: State.DIVIDE_ASSIGN,
    },
    [State.ASSIGN]: {
      isEndable: true,
      tokenType: State.ASSIGN,
    },
    [State.LESS_THAN]: {
      isEndable: true,
      tokenType: State.LESS_THAN,
      transition: [
        {
          state: State.LESS_EQUAL_THAN,
          checker: '='
        }
      ]
    },
    [State.LESS_EQUAL_THAN]: {
      isEndable: true,
      tokenType: State.LESS_EQUAL_THAN,
    },
    [State.GREATER_THAN]: {
      isEndable: true,
      tokenType: State.GREATER_THAN,
      transition: [
        {
          state: State.GREATER_EQUAL_THAN,
          checker: '='
        }
      ]
    },
    [State.GREATER_EQUAL_THAN]: {
      isEndable: true,
      tokenType: State.GREATER_EQUAL_THAN,
    },
    [State.LINE_COMMENT]: {
      isEndable: true,
      tokenType: State.LINE_COMMENT,
      transition: [
        {
          state: State.LINE_COMMENT,
          checker: (char: string) => !NEW_LINE.includes(char)
        }
      ]
    },
    [State.START_BLOCK_COMMENT]: {
      isEndable: false,
      transition: [
        {
          state: State.BLOCK_COMMENT,
          checker: (char, buffer) => /^\/\*[\d\D]+\*\/$/.test(buffer + char),
        },
        {
          state: State.START_BLOCK_COMMENT,
          checker: (char, buffer) => /^\/\*/.test(buffer + char)
        },
      ]
    },
    [State.BLOCK_COMMENT]: {
      isEndable: true,
      tokenType: State.BLOCK_COMMENT
    }
  }
}

export default config