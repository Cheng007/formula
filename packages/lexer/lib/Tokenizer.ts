import TokenBuffer from "./TokenBuffer";
import type { IStatesConfig, IStateKey, IConfig, ITokenTypeGenerator, IToken, TOKEN_TYPE } from "./types";
import Position from './Position'
import { NEW_LINE, SPACE } from "./config";

export default class Tokenizer {
  #tokenBuffer = new TokenBuffer()
  #statesConfig: IStatesConfig
  #initialState: IStateKey
  #state: IStateKey
  #position = new Position()
  #buffer = ''

  constructor(config: IConfig) {
    this.#statesConfig = config.states
    this.#initialState = config.initialState
    this.#state = config.initialState
  }

  getTokenBuffer() {
    return this.#tokenBuffer
  }

  parse(code: string) {
    for (let i = 0; i < code.length; i++) {
      const char = code[i]
      this.consume(char)
    }
    this.end()

    return this.#tokenBuffer.getTokens()
  }

  syntaxError(errorMsg: string) {
    const { line, column } = this.#position.getCurrentPosition()
    const msg = `${errorMsg} at line ${line}, column ${column}`
    return new SyntaxError(msg)
  }

  end() {
    if (this.#state === this.#initialState) {
      if (this.#buffer) {
        throw this.syntaxError('Unexpected EOF')
      }
      return
    }

    const currentStateConfig = this.#statesConfig[this.#state]

    if (currentStateConfig.isEndable && currentStateConfig.tokenType) {
      this.addToken(currentStateConfig.tokenType)
    }
  }

  addToken(tokeType: TOKEN_TYPE | ITokenTypeGenerator) {
    const startLocation = this.#position.getPrevLocation()
    const endLocation = this.#position.getCurrentPosition()
    const token: IToken = {
      type: typeof tokeType === 'function' ? tokeType(this.#buffer) : tokeType,
      value: this.#buffer,
      range: {
        start: startLocation!,
        end: endLocation
      }
    }
    this.#tokenBuffer.write(token)
  }

  reset() {
    this.#state = this.#initialState
    this.#buffer = ''
  }

  consume(char: string) {
    if (
      (char === SPACE || NEW_LINE.includes(char))
      && this.#state === this.#initialState
    ) {
      this.#position.move(char)
      return
    }

    const currentStateConfig = this.#statesConfig[this.#state]
    if (!currentStateConfig) {
      throw new Error(`Missing state config for ${this.#state}`)
    }

    const transition = currentStateConfig.transition
    if (!transition) {
      if (currentStateConfig.isEndable) {
        if (currentStateConfig.tokenType) {
          this.addToken(currentStateConfig.tokenType)
        }
        this.reset()
        this.consume(char)
        return
      }

      throw this.syntaxError(`Unexpected character ${char}`)
    }

    const targetTransition = transition.find(({ checker }) => {
      if (typeof checker === 'string') {
        return char === checker
      }

      if (checker instanceof RegExp) {
        return checker.test(char)
      }

      return checker(char, this.#buffer)
    })

    if (!targetTransition) {
      if (currentStateConfig.isEndable) {
        if (currentStateConfig.tokenType) {
          this.addToken(currentStateConfig.tokenType)
        }
        this.reset()
        this.consume(char)
        return
      }

      this.#position.move(char)
      throw this.syntaxError(`Invalid or unexpected token ${char}`)
    }

    this.#position.move(char)
    // when state start to transfer from initial state to other state, mark current position
    if (
      this.#state === this.#initialState
      && targetTransition?.state !== this.#initialState
    ) {
      this.#position.markPosition()
    }

    this.#state = targetTransition!.state
    this.#buffer += char
  }
}