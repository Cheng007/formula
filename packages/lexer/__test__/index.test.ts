import Tokenizer from "../lib/Tokenizer";
import config from '../lib/config';

// test for 行注释、块注释

const code = `
  /* 放大 */fda
  /*
  fa
  sfds*/fdsa
`

const testCases: Array<{ code: string, description: string }> = [
  {
    code: `abc`,
    description: 'it should be identifier'
  },
  {
    code: `'abc'`,
    description: 'it should be string'
  },
  {
    code: `1234.45`,
    description: 'it should be number'
  },
  {
    code: `
      /* 行内注释 */
      /*
      块级注释
      */
    `,
    description: 'it should be comment'
  }
]

describe('Token with config', () => {
  testCases.forEach(({ code, description }) => {
    test(description, () => {
      const tokenizer = new Tokenizer(config)
      tokenizer.parse(code)
      const tokenBuffer = tokenizer.getTokenBuffer()
      expect(tokenBuffer.getTokens()).toMatchSnapshot()
    })
  })
})
