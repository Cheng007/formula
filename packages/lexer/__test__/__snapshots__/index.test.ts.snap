// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Token with config it should be comment 1`] = `
[
  {
    "range": {
      "end": {
        "column": 16,
        "line": 2,
      },
      "start": {
        "column": 7,
        "line": 2,
      },
    },
    "type": "BLOCK_COMMENT",
    "value": "/* 行内注释 */",
  },
  {
    "range": {
      "end": {
        "column": 8,
        "line": 5,
      },
      "start": {
        "column": 7,
        "line": 3,
      },
    },
    "type": "BLOCK_COMMENT",
    "value": "/*
      块级注释
      */",
  },
]
`;

exports[`Token with config it should be identifier 1`] = `
[
  {
    "range": {
      "end": {
        "column": 3,
        "line": 1,
      },
      "start": {
        "column": 1,
        "line": 1,
      },
    },
    "type": "IDENTIFIER",
    "value": "abc",
  },
]
`;

exports[`Token with config it should be number 1`] = `
[
  {
    "range": {
      "end": {
        "column": 7,
        "line": 1,
      },
      "start": {
        "column": 1,
        "line": 1,
      },
    },
    "type": "NUMBER",
    "value": "1234.45",
  },
]
`;

exports[`Token with config it should be string 1`] = `
[
  {
    "range": {
      "end": {
        "column": 5,
        "line": 1,
      },
      "start": {
        "column": 1,
        "line": 1,
      },
    },
    "type": "STRING",
    "value": "'abc'",
  },
]
`;
